static/js/recent_topics_ui.js

```diff
@@ -274,6 +274,7 @@ function format_topic(topic_data) {
     const stream_muted = stream_data.is_muted(stream_id);
     const muted = topic_muted || stream_muted;
     const unread_count = unread.num_unread_for_topic(stream_id, topic);
+    const mention_in_unread = unread.stream_has_any_unread_mentions(stream_id) && unread_count;
```

Comment:
This just checks whether the stream has unread mentions and the topic has unread messages, not that those are the same messages!

```
export function stream_has_any_unread_mentions(stream_id) {                                              
    const streams_with_mentions = unread_topic_counter.get_streams_with_unread_mentions();               
    return streams_with_mentions.has(stream_id);                                                         
}                                                                                                        
```

there's also a performance issue; the way that these functions work is that we loop through all unread mentions and look up the stream/topic they're in. As a result, it's the same work to return __all__ stream/topic pairs with unread mentions as it will be to ask if a single topic has unread mentions; linear in the total number of mentions. So rendering Recent Topics with 100 unread mentions and 1000 rows in Recent Topics would be O(100K) work if we call that function for every topic, which is likely to take several seconds.

What we do to work around this in `topic_list_data.js` is do a single bulk query for the topic; we could readily write a similar one that did a single bulk query for everything:

```
    get_topics_with_unread_mentions(stream_id) {                                                         
        // Returns the set of lower cased topics with unread mentions                                    
        // in the given stream.                                                                          
        const result = new Set();                                                                        
        const per_stream_bucketer = this.bucketer.get_bucket(stream_id);                                 
                                                                                                         
        if (!per_stream_bucketer) {                                                                      
            return result;                                                                               
        }                                                                                                
                                                                                                         
        for (const message_id of unread_mentions_counter) {                                              
            // Because bucket keys in per_stream_bucketer are topics,                                    
            // we can just directly use reverse_lookup to find the                                       
            // topic in this stream containing a given unread message                                    
            // ID. If it's not in this stream, we'll get undefined.                                      
            const topic_match = per_stream_bucketer.reverse_lookup.get(message_id);                      
            if (topic_match !== undefined) {                                                             
                // Important: We lower-case topics here before adding them                               
                // to this set, to support case-insensitive checks.                                      
                result.add(topic_match.toLowerCase());                                                   
            }                                                                                            
        }                                                                                                
                                                                                                         
        return result;                                                                                   
    }                                                                                                    
```

Reply:
I also don't know whether 1000 Recent Topics rows being rendered by a `list_widget` is a realistic scenario; it's possible we can decide it's fast enough even with the quadratic algorithm ... but I'd much prefer avoiding that since performance issues can make a big impression, and this is meant to be a home view that it's fast to switch to.